mne/export/_edf.py

```diff
@@ -0,0 +1,184 @@
+# -*- coding: utf-8 -*-
+# Authors: MNE Developers
+#
+# License: BSD-3-Clause
+
+import numpy as np
+
+from ..utils import _check_edflib_installed, warn
+_check_edflib_installed()
+from EDFlib.edfwriter import EDFwriter  # noqa: E402
+
+
+def _export_raw(fname, raw, physical_range, fmt):
+    phys_dims = 'uV'
+
+    if fmt == 'bdf':
+        digital_min = -8388607
+        digital_max = 8388607
+    elif fmt == 'edf':
+        digital_min = -32767
+        digital_max = 32767
+
+    # load data first
+    raw.load_data()
+
+    # remove extra epoc and STI channels
+    drop_chs = ['epoc']
+    if not (raw.filenames[0].endswith('.fif')):
+        drop_chs.append('STI 014')
+
+    ch_names = [ch for ch in raw.ch_names if ch not in drop_chs]
+    n_chs = len(ch_names)
+    n_times = raw.n_times
+    sfreq = int(raw.info['sfreq'])
+    n_secs = n_times / sfreq
+
+    # get data in uV
+    units = dict()
+    if 'eeg' in raw:
+        units['eeg'] = 'uV'
+    if 'ecog' in raw:
+        units['ecog'] = 'uV'
+    if 'seeg' in raw:
+        units['seeg'] = 'uV'
+
+    # get the entire dataset in uV
+    data = raw.get_data(units=units, picks=ch_names)
+
+    if physical_range == 'auto':
+        # get max and min for each channel type data
+        ch_types_phys_max = dict()
+        ch_types_phys_min = dict()
+        for ch_type in raw.get_channel_types():
+            ch_type_data = raw.copy().pick_types(
+                **{ch_type: True}).get_data(units=units)
+            ch_types_phys_max[ch_type] = ch_type_data.max()
+            ch_types_phys_min[ch_type] = ch_type_data.min()
+    else:
+        # get the physical min and max of the data in uV
+        # see discussion in: https://github.com/sccn/eeglab/issues/246
+        pmin, pmax = -3200, 3200
+
+        # check that physical min and max is not exceeded
+        if data.max() > pmax:
+            raise RuntimeError('The maximum uV of the data is more then pmax.')
+        if data.min() < pmin:
+            raise RuntimeError('The minimum uV of the data is less then pmin.')
+
+    # create instance of EDF Writer
+    hdl = EDFwriter(fname, EDFwriter.EDFLIB_FILETYPE_EDFPLUS, n_chs)
+
+    # set channel data
+    for ichan, ch in enumerate(ch_names):
+        if physical_range == 'auto':
+            # take the channel type minimum and maximum
+            ch_type = raw.get_channel_types(picks=ch)[0]
+            pmin, pmax = ch_types_phys_min[ch_type], ch_types_phys_max[ch_type]
+
+        if hdl.setPhysicalMaximum(ichan, pmax) != 0:  # noqa
```

Comment:
Removed.

Reply:
So the issue is that these set functions are hard to test for the errors since they are a consequence of just the data being formatted incorrectly. But I removed them for now.